src/main/java/picard/sam/SplitSamBySize.java

```diff
@@ -0,0 +1,152 @@
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SplitSamBySize \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000 \\ <br />" +
+            "     SPLIT_TO_N_READS=48000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N_READS", doc = "Split to have approximately N reads per output file.", optional = true,
+            mutex = {"SPLIT_TO_N_FILES"})
+    public int SPLIT_TO_N_READS;
+
+    @Argument(shortName = "N_FILES", doc = "Split to N files.", optional = true, mutex = {"SPLIT_TO_N_READS"})
+    public int SPLIT_TO_N_FILES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = "O", doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT_DIR;
+
+    @Argument(shortName = "OUT_PREFIX", doc = "Prefix to name output files.")
+    public String OUT_PREFIX = "shard";
+
+    private final Log log = Log.getInstance(SplitSamBySize.class);
+
+    public static void main(final String[] argv) {
+        System.exit(new SplitSamBySize().instanceMain(argv));
+    }
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+        final SAMFileHeader header = reader.getFileHeader();
+        if (header.getSortOrder() == SAMFileHeader.SortOrder.coordinate) {
+            log.warn("Splitting a coordinate sorted bam may result in invalid bams " +
+                    "that do not always contain each read's mate in the same bam.");
+        }
+
+        final SAMFileWriterFactory factory = new SAMFileWriterFactory();
+
+        int splitToNFiles = SPLIT_TO_N_FILES != 0 ? SPLIT_TO_N_FILES : (int) Math.ceil(TOTAL_READS_IN_INPUT / SPLIT_TO_N_READS);
+        int readsPerFile = (int) Math.ceil(TOTAL_READS_IN_INPUT / (double) splitToNFiles);
+        SAMFileWriter[] writers = generateWriters(factory, header, splitToNFiles);
+
+        int recordsWritten = 0;
+        int writerIndex = 0;
+        String lastReadName = "";
+        final ProgressLogger progress = new ProgressLogger(log);
+        for (final SAMRecord currentRecord : reader) {
+            if (recordsWritten >= readsPerFile && !lastReadName.equals(currentRecord.getReadName())) {
+                writerIndex++;
+                recordsWritten = 0;
+            }
+            writers[writerIndex].addAlignment(currentRecord);
+            lastReadName = currentRecord.getReadName();
+            recordsWritten++;
+            progress.record(currentRecord);
+        }
```

Comment:
Would you like to emit a warning if the number of reads found differs from the number of reads predicted?